lib/assert.js

```diff
@@ -931,6 +933,78 @@ assert.doesNotMatch = function doesNotMatch(string, regexp, message) {
   internalMatch(string, regexp, message, doesNotMatch);
 };
 
+const noop = () => {};
+
+const mustCallChecks = [];
+
+let attachExitHandler = true;
+
+function runCallChecks(exitCode) {
+  for (const context of mustCallChecks) {
+    if (context.actual !== context.exact) {
+      context.error = `Mismatched ${context.name} function calls. ` +
+        `Expected exactly ${context.exact}, actual ${context.actual}.`;
+      const console = require('console');
+      console.error(context.error);
+      console.error(context.stackTrace);
+      if (exitCode === 0)
+        process.exitCode = 1;
+    }
+  }
+}
+
+assert.mustCall = function mustCall(fn, exact = 1) {
+  if (process._exiting)
+    throw new ERR_MUSTCALL_UNAVAILABLE_DURING_EXIT();
+  if (typeof fn === 'number') {
+    exact = fn;
+    fn = noop;
+  } else if (fn === undefined) {
+    fn = noop;
+  }
+
+  if (!Number.isInteger(exact) || exact < 1)
+    throw new ERR_INVALID_ARG_TYPE(`Invalid exact value: ${exact}`);
+
+  const context = {
+    exact,
+    actual: 0,
+    // eslint-disable-next-line no-restricted-syntax
+    stackTrace: new Error(),
+    name: fn.name || 'mustCall'
+  };
+
+  // Add the exit listener only once to avoid listener leak warnings
+  if (attachExitHandler === true) {
+    attachExitHandler = false;
+    process.on('exit', runCallChecks);
+  }
+
+  mustCallChecks.push(context);
```

Comment:
When switching to a set:

```suggestion
  mustCallChecks.add(context);
```